[Dashboard] Focus on a single panel

cqliu1 opened the pull request:
## Summary

Closes #163274.

This adds support for focusing a single panel on a dashboard while blurring/disabling all other panels.

https://github.com/elastic/kibana/assets/1697105/2e45b2fc-14e0-4cf4-8b83-1c07187babc9

All interactions on blurred panels are disabled as well as any layout changes on either the focused or blurred panels, but the user still has access to controls, the editor toolbar, and unified search.

**Note:** For testing purposes, I have the focus behavior implemented with the `Panel settings` action, but I won't merge this PR with focus enabled for `Panel settings`. This behavior is meant for flyouts with inline editing capabilities, like the future ESQL inline editor flyout where changes apply in real-time without hitting `Apply` to see the changes.

### Checklist

Delete any items that are not applicable to this PR.

- [ ] Any text added follows [EUI's writing guidelines](https://elastic.github.io/eui/#/guidelines/writing), uses sentence case text and includes [i18n support](https://github.com/elastic/kibana/blob/main/packages/kbn-i18n/README.md)
- [ ] [Documentation](https://www.elastic.co/guide/en/kibana/master/development-documentation.html) was added for features that require explanation or tutorials
- [ ] [Unit or functional tests](https://www.elastic.co/guide/en/kibana/master/development-tests.html) were updated or added to match the most common scenarios
- [ ] Any UI touched in this PR is usable by keyboard only (learn more about [keyboard accessibility](https://webaim.org/techniques/keyboard/))
- [ ] Any UI touched in this PR does not create any new axe failures (run axe in browser: [FF](https://addons.mozilla.org/en-US/firefox/addon/axe-devtools/), [Chrome](https://chrome.google.com/webstore/detail/axe-web-accessibility-tes/lhdoppojpmngadmnindnejefpokejbdd?hl=en-US))
- [ ] If a plugin configuration key changed, check if it needs to be allowlisted in the cloud and added to the [docker list](https://github.com/elastic/kibana/blob/main/src/dev/build/tasks/os_packages/docker_generator/resources/base/bin/kibana-docker)
- [ ] This renders correctly on smaller devices using a responsive layout. (You can test this [in your browser](https://www.browserstack.com/guide/responsive-testing-on-local-server))
- [ ] This was checked for [cross-browser compatibility](https://www.elastic.co/support/matrix#matrix_browsers)


### Risk Matrix

Delete this section if it is not applicable to this PR.

Before closing this PR, invite QA, stakeholders, and other developers to identify risks that should be tested prior to the change/feature release.

When forming the risk matrix, consider some of the following examples and how they may potentially impact the change:

| Risk                      | Probability | Severity | Mitigation/Notes        |
|---------------------------|-------------|----------|-------------------------|
| Multiple Spaces&mdash;unexpected behavior in non-default Kibana Space. | Low | High | Integration tests will verify that all features are still supported in non-default Kibana Space and when user switches between spaces. |
| Multiple nodes&mdash;Elasticsearch polling might have race conditions when multiple Kibana nodes are polling for the same tasks. | High | Low | Tasks are idempotent, so executing them multiple times will not result in logical error, but will degrade performance. To test for this case we add plenty of unit tests around this logic and document manual testing procedure. |
| Code should gracefully handle cases when feature X or plugin Y are disabled. | Medium | High | Unit tests will verify that any feature flag or plugin combination still results in our service operational. |
| [See more potential risk examples](https://github.com/elastic/kibana/blob/main/RISK_MATRIX.mdx) |


### For maintainers

- [ ] This was checked for breaking API changes and was [labeled appropriately](https://www.elastic.co/guide/en/kibana/master/contributing.html#kibana-release-notes-process)

cqliu1 committed:
Focus on a single panel

cqliu1 commented:
Do we want to disable controls, editor toolbar, and/or unified search access when focused on a single panel as well?
@teresaalvarezsoler @andreadelrio @stratoula

cqliu1 committed:
Add focusPanelId option to openOverlay

teresaalvarezsoler commented:
> Do we want to disable controls, editor toolbar, and/or unified search access when focused on a single panel as well?

When editing the panel settings it doesn't make much sense to filter visualizations but they don't hurt either so I would leave them accessible for now. In the future, we can reopen the discussion when users are able to edit panel-level filters from there.

Accessing the filters, controls, etc. definetely makes more sense when users are editing a visualization, @stratoula is it fine for ESQL visualizations? I think this doesn't work in Discover now.

stratoula commented:
@teresaalvarezsoler not sure I have understood the question correctly but I will give it a shot. We have disabled the kql bar and the filter builder in Discover as you can do all these things with ES|QL. On a dashboard, where we don't have the ES|QL editor we have left the kql bar/filter builder which works fine with dataview and ESQL charts.

teresaalvarezsoler commented:
Great, then let's leave them accessible @cqliu1

cqliu1 committed:
Remove hover background and cursor on emb panel when focused

cqliu1 committed:
Use emotion for focus styles

cqliu1 commented:
RE: a11y, it looks like even though mouse events are disabled for the blurred panels, you can still tab over to the buttons in the blurred panels before reaching the flyout. Do we want to prevent keyboard navigation to the blurred panels? If so, is this a blocker, or can we improve a11y in a follow up PR?

@teresaalvarezsoler @andreadelrio

cqliu1 committed:
Remove unused prop

elasticmachine commented:
Pinging @elastic/kibana-presentation (Team:Presentation)

stratoula commented:
@cqliu1 this looks great! Can we make it also blur when the user clicks the edit Visualization action on an ESQL chart?

<img width="1247" alt="image" src="https://github.com/elastic/kibana/assets/17003240/a46210e9-4c3d-4ae2-9196-f80f76e75e09">

To create an ESQL chart, go to Discover, click the dataview picker and Try ES|QL. Try this query
`from kibana_sample_data_logs | limit 10`, save the chart to a dashboard (using the save icon which is next to the chart)

<img width="2484" alt="image" src="https://github.com/elastic/kibana/assets/17003240/642ba3a6-a00b-49ba-bbf8-6b8fa97da695">

teresaalvarezsoler commented:
> Do we want to prevent keyboard navigation to the blurred panels? If so, is this a blocker, or can we improve a11y in a follow up PR?

@cqliu1 I think we can open a new issue as a follow-up improvement for later

cqliu1 committed:
Blur panels on open lens config action instead of customize panel action

cqliu1 committed:
Remove unused import

cqliu1 commented:
I've updated the PR to blur panels for the `Edit ESQL visualizations` action and removed the blurring for the `Panel settings` action.

<img width="1800" alt="Screenshot 2023-09-18 at 10 46 39 AM" src="https://github.com/elastic/kibana/assets/1697105/ad59d987-ead4-41b8-b4f7-65e5e05ba3f1">

stratoula commented:
You are a gem @cqliu1 ❤️

andreadelrio reviewed (changes requested):
Looking good! The interactions of this (opening/closing the push flyout, rearrangement of the dashboard layout) feel a bit slow and not smooth to me. I wonder if there's a way to optimize this?

When you open a flyout within a flyout the left side border looks too thick and with an extra shadow. We need to adjust it so it looks the same even there's a flyout on top of a flyout and when there's only one flyout. See:

#### One flyout
<img width="478" alt="image" src="https://github.com/elastic/kibana/assets/4016496/e6cf263c-c28a-4435-bc7e-d659aacefd72">

#### Flyout on top of flyout
<img width="412" alt="image" src="https://github.com/elastic/kibana/assets/4016496/abfec14b-e7de-4f0b-b773-2191f5d0c631">

stratoula reviewed (approved):
This looks great I love it, visualizations team changes LGTM!

@andreadelrio the shadow thing you mention seems a bit irrelevant with this PR(?) right?

MichaelMarcialis commented:
> > Do we want to disable controls, editor toolbar, and/or unified search access when focused on a single panel as well?
> 
> When editing the panel settings it doesn't make much sense to filter visualizations but they don't hurt either so I would leave them accessible for now. In the future, we can reopen the discussion when users are able to edit panel-level filters from there.
> 
> Accessing the filters, controls, etc. definetely makes more sense when users are editing a visualization, @stratoula is it fine for ESQL visualizations? I think this doesn't work in Discover now.

@cqliu1, @teresaalvarezsoler, @stratoula: [Per the designs](https://www.figma.com/file/s62do8GDjERrRtp0sKKCJy/Mockups-%2F-Version-1-%5BIn-Progress%5D?type=design&node-id=277%3A267965&mode=design&t=vNHh8aZbqxtm9Bf7-1), our suggestion was to leave the unified search interface enabled/accessible, but to disable the editor toolbar when a single panel is focused. Regarding custom controls, I don't think we made a decision regarding whether to enable/disable, so I'll leave that decision up to ya'll (I can see arguments for on both sides).

> RE: a11y, it looks like even though mouse events are disabled for the blurred panels, you can still tab over to the buttons in the blurred panels before reaching the flyout. Do we want to prevent keyboard navigation to the blurred panels? If so, is this a blocker, or can we improve a11y in a follow up PR?

@cqliu1, @teresaalvarezsoler: Ideally we'd prevent keyboard navigation to those items that are being given a visually disabled treatment from the initial release of this feature, to avoid users inadvertently cancelling out of one workflow and being thrown into another. But whether this should be part of this PR or a separate one, I'll leave up to ya'll to decide.

> When you open a flyout within a flyout the left side border looks too thick and with an extra shadow. We need to adjust it so it looks the same even there's a flyout on top of a flyout and when there's only one flyout.

@andreadelrio: I think the 2px wide border is an intentional style coming from EUI (specifically for push flyouts). I'm personally not a fan of it either. If you agree, we can probably open an EUI issue to change it. 

That aside, I agree with you on the shadow for the one-level-deeper (i.e. dimension item) flyout experience. We don't want to give the user the impression that another flyout is being rendered on top of the root-level push flyout. Instead, we want to give the impression that the contents of the root-level flyout are being replaced with the dimension item configuration. @stratoula, is that something we're able to adjust as part of your ES|QL work (since it's probably not directly part of this PR).

stratoula commented:
@MichaelMarcialis about that

> @stratoula, is that something we're able to adjust as part of your ES|QL work (since it's probably not directly part of this PR)

Absolutely but I have to admit that I dont understand exactly the request 🙈

MichaelMarcialis commented:
> @MichaelMarcialis about that
> 
> > @stratoula, is that something we're able to adjust as part of your ES|QL work (since it's probably not directly part of this PR)
> 
> Absolutely but I have to admit that I dont understand exactly the request 🙈

Haha, no worries. Forgive my poor description. I'll throw a time on the calendar for us to discuss via Zoom.

stratoula commented:
Thank you Michael! Please do ❤️

cqliu1 commented on src/plugins/dashboard/public/dashboard_container/component/grid/dashboard_grid_item.tsx:
```diff
@@ -22,7 +23,6 @@ export interface Props extends DivProps {
   id: DashboardPanelState['explicitInput']['id'];
   index?: number;
   type: DashboardPanelState['type'];
-  focusedPanelId?: string;
```
@ThomThomson This prop doesn't actually seem to be used anywhere, so I removed it to reduce confusion.

ThomThomson replied:
Thanks @cqliu1! I think this was something that was introduced long ago _anticipating_ a focused panel, but the feature didn't exist.

ThomThomson commented:
@andreadelrio 
> Looking good! The interactions of this (opening/closing the push flyout, rearrangement of the dashboard layout) feel a bit slow and not smooth to me. I wonder if there's a way to optimize this?

I've noticed that for certain Dashboards in main and in this PR, that sometimes the Panels don't animate on window size changes. I'm also looking into a way to trigger the size change programatically, so we can ensure it resizes as soon as the push flyout is opened.

cqliu1 committed:
Disable editor toolbar while panel focused

cqliu1 commented:
> @cqliu1, @teresaalvarezsoler, @stratoula: [Per the designs](https://www.figma.com/file/s62do8GDjERrRtp0sKKCJy/Mockups-%2F-Version-1-%5BIn-Progress%5D?type=design&node-id=277%3A267965&mode=design&t=vNHh8aZbqxtm9Bf7-1), our suggestion was to leave the unified search interface enabled/accessible, but to disable the editor toolbar when a single panel is focused. Regarding custom controls, I don't think we made a decision regarding whether to enable/disable, so I'll leave that decision up to ya'll (I can see arguments for on both sides).

I've disabled the editor toolbar while a panel is focused, but I'll leave controls accessible.

<img width="1800" alt="Screenshot 2023-09-20 at 8 00 33 AM" src="https://github.com/elastic/kibana/assets/1697105/6e26cea4-3a5d-495d-8f29-8e1149a903d9">

cqliu1 committed:
Remove console log

ThomThomson commented:
I made some small changes to the debounce effect on width changes:
1. I changed it from 250 ms to 100ms,
2. I skip the debounce entirely when the Dashboard has an overlay open.

This results in a smoother open / close experience.

![skip debounce](https://github.com/elastic/kibana/assets/14276393/b7e0a271-f9f0-437d-b225-3741a31e4dc8)

cqliu1 committed:
Skip debounce when overlays exist

cqliu1 committed:
Merge branch 'main' of https://github.com/elastic/kibana into dashboard/focus-panel

cqliu1 commented:
> Looking good! The interactions of this (opening/closing the push flyout, rearrangement of the dashboard layout) feel a bit slow and not smooth to me. I wonder if there's a way to optimize this?

@andreadelrio With Devon's changes that remove the debounce from overlays, opening and closing flyouts looks much smoother for the ESQL and panel settings flyouts.

![Sep-20-2023 10-38-38](https://github.com/elastic/kibana/assets/1697105/60f525e6-f019-473c-a4f9-66ddfde4fea4)

cqliu1 committed:
Restore panel settings flyout back to overlay instead of push

cqliu1 committed:
Remove blurred items from tab order

cqliu1 commented:
In https://github.com/elastic/kibana/pull/165417/commits/3f02e1ad7d7b0fc34d433901208964d4a7be0a3e, I figured out a way to programmatically remove blurred panels and nested elements from the tab order to disable keyboard navigation.


https://github.com/elastic/kibana/assets/1697105/1cf76b51-5681-4c65-ad6f-b27dc929714d

cqliu1 commented:
@elastic/appex-sharedux I'm not sure why you weren't listed as reviewer, but I enabled the `isDisabled` prop in the toolbar button so I could disable the editor toolbar when the ESQL flyout is open.

ThomThomson commented on src/plugins/dashboard/public/dashboard_app/top_nav/controls_toolbar_button/controls_toolbar_button.tsx:
```diff
@@ -17,7 +17,12 @@ import { AddDataControlButton } from './add_data_control_button';
+  isDisabled?: boolean;
+}
+
+export const ControlsToolbarButton: FC<Props> = ({ controlGroup, isDisabled }) => {
```
nit: why the use of FC here? Using FC like this [implicitly adds](https://medium.com/raccoons-group/why-you-probably-shouldnt-use-react-fc-to-type-your-react-components-37ca1243dd13) the `children` prop, and this toolbar button doesn't actually take children. It's also a little cleaner to just declare the Props type inline instead of naming it.

ThomThomson commented on src/plugins/dashboard/public/dashboard_container/component/grid/dashboard_grid_item.tsx:
```diff
@@ -43,20 +42,24 @@ const Item = React.forwardRef<HTMLDivElement, Props>(
       // https://github.com/react-grid-layout/react-grid-layout/issues/1241#issuecomment-658306889
       children,
       className,
-      style,
```
What was style used for? Nothing?

ThomThomson reviewed (approved):
Changes LGTM! Tested this in chrome, and everything looks good (especially the tab ordering). Also looked through the changes, and left one nit.

Additionally, can you think of a good way to test this, preferably with unit tests? Maybe checking for the presence of a class or a data test subj when focus panel is defined?

sebelga reviewed (approved):
Shared ux changes LGTM 👍

cqliu1 commented on src/plugins/dashboard/public/dashboard_container/component/grid/dashboard_grid_item.tsx:
```diff
@@ -69,12 +72,30 @@ const Item = React.forwardRef<HTMLDivElement, Props>(
 
     return (
       <div
-        style={{ ...style, zIndex: focusedPanelId === id ? 2 : 'auto' }}
```
@ThomThomson Before `style` was destructured from props to override z-index on `focusedPanelId` which wasn't being used, and it doesn't seem necessary to change the z-index, so I removed the z-index style and removed the `style` previously destructured on line 46.

andreadelrio reviewed (approved):
LGTM, thanks for the updates. Nice job on this!

cqliu1 committed:
Skip debounce only if panel is focused

cqliu1 committed:
Undo FC typing

kibanamachine committed:
[CI] Auto-commit changed files from 'node scripts/precommit_hook.js --ref HEAD~1..HEAD --fix'

gchaps commented:
Can the menu item start with a verb, for example "Apply settings", so that it follows the format of the other items in the menu?

cqliu1 committed:
Rename var and added dashboard grid unit tests

cqliu1 committed:
Merge branch 'dashboard/focus-panel' of https://github.com/cqliu1/kibana into dashboard/focus-panel

cqliu1 committed:
Merge branch 'main' of https://github.com/elastic/kibana into dashboard/focus-panel

gchaps commented:
After talking with @cqliu1, we decided on `Change panel settings`.

ThomThomson commented:
> we decided on Change panel settings.
@gchaps @cqliu1  this item also exists in view mode, and in view mode you can't really change the settings. If we _must_ have a verb, maybe we should switch it depending on the view mode?

cqliu1 commented:
Yeah in edit mode you can change title, description, tags, along with the time range, but in view mode, you can only edit time range.

#### Edit mode
<img width="419" alt="Screenshot 2023-09-25 at 12 24 23 PM" src="https://github.com/elastic/kibana/assets/1697105/e2db0897-9a40-4f00-80b2-8b7acc7e4a88">

#### View mode
<img width="439" alt="Screenshot 2023-09-25 at 12 24 04 PM" src="https://github.com/elastic/kibana/assets/1697105/cc18c171-01c0-4950-a3d4-5199882185fc">

Let's just leave it as `Panel settings`.

cqliu1 committed:
Fix jest test

cqliu1 committed:
Add grid item unit tests
